### PR TITLE
Better user experience for using 2FA in sonata

### PR DESCRIPTION
**Pull request by aLkRicha**
I am targeting this branch, because there is no BC changes.

## Changelog

```markdown
### Added
- New experience for `sonata-project/google-authenticator` users. Showing QR-code in login form, automatically setting 2FA secret to user.
```
## Subject
Better user experience for using 2FA in sonata
Imgur
Imgur

### CODE REVIEW

**Comment by greg0ire** [comment]
Please write a changelog directed at end users. Keep in mind that they probably do not care what file or class you changed, but what features you provided / what bugs you fixed. Try to answer the question "What benefit do I get from upgrading?"
More on http://keepachangelog.com

File: docs/reference/two_step_validation.rst [hunk 1]
```diff
@@ -35,5 +35,38 @@ Edit the configuration file:
         google_authenticator:
             enabled: true
             server:  yourserver.com
+            ip_white_list:
+                - 127.0.0.1
+            forced_for_role:
+                - ROLE_ADMIN
+
+Also, if you want to use ``ip_white_list`` and ``forced_for_role`` configuration for automatically setting secret
```
Comment (greg0ire) [inline docs/reference/two_step_validation.rst]: - "configuration nodes"
- "the secret" ? (should be singular, it's a secret, right?)

File: docs/reference/two_step_validation.rst [hunk 2]
```diff
@@ -35,5 +35,38 @@ Edit the configuration file:
...
+to user and showing QR-code in login form, you need to set success handler in your firewall to ``sonata.user.google.authenticator.success_handler``, example:
```
Comment (greg0ire) [inline docs/reference/two_step_validation.rst]: What does "setting secret to user" mean here exactly? I'm a noob.

Comment (greg0ire) [reply docs/reference/two_step_validation.rst]: "to set the success handler"

File: docs/reference/two_step_validation.rst [hunk 3]
```diff
@@ -35,5 +35,38 @@ Edit the configuration file:
...
+                form_login:
+                    provider:        fos_userbundle
+                    login_path:      /admin/login
+                    use_forward:     false
+                    check_path:      /admin/login_check
+                    failure_path:    null
+                    success_handler: sonata.user.google.authenticator.success_handler
+
+
+Then after success login, if user need to use 2FA and has no secret, qr-code will be shown on login form.
```
Comment (greg0ire) [inline docs/reference/two_step_validation.rst]: - "if the user needs to use"
- "a QR code"
- "in the login form"

File: src/DependencyInjection/Configuration.php [hunk 4]
```diff
@@ -61,6 +61,14 @@ public function getConfigTreeBuilder()
                     ->children()
                         ->scalarNode('server')->cannotBeEmpty()->end()
                         ->scalarNode('enabled')->defaultFalse()->end()
+                        ->arrayNode('ip_white_list')
+                            ->prototype('scalar')->end()
+                            ->defaultValue(['127.0.0.1'])
```
Comment (greg0ire) [inline src/DependencyInjection/Configuration.php]: Please add calls to `info`

File: docs/reference/two_step_validation.rst (hunk 2 above)
Comment (aLkRicha) [reply docs/reference/two_step_validation.rst]: @greg0ire the secret
to user (secret - a connection between user and device that will scans QR-code)

File: docs/reference/two_step_validation.rst [hunk 5]
```diff
@@ -40,8 +40,8 @@ Edit the configuration file:
             forced_for_role:
                 - ROLE_ADMIN
 
-Also, if you want to use ``ip_white_list`` and ``forced_for_role`` configuration for automatically setting secret
-to user and showing QR-code in login form, you need to set success handler in your firewall to ``sonata.user.google.authenticator.success_handler``, example:
+Also, if you want to use ``ip_white_list`` and ``forced_for_role`` configuration nodes for automatically setting the secret
+to user (secret - a connection between user and device that will scans QR-code) and showing QR-code in login form, you need to set the success handler in your firewall to ``sonata.user.google.authenticator.success_handler``, example:
```
Comment (greg0ire) [inline docs/reference/two_step_validation.rst]: "will scan"

Comment (greg0ire) [reply docs/reference/two_step_validation.rst]: Please comply with the sf docs standards:

> [Each line should break approximately after the first word that crosses the 72nd character (so most lines end up being 72-78 characters);](http://symfony.com/doc/current/contributing/documentation/standards.html)

**Comment by greg0ire** [comment]
Please kindly squash your commits together. If you don't, we'll try to remember to do it for you but it's best if you save us this trouble.

How to do that?

1. `git rebase -i origin/3.x`, assuming `origin` is a git remote that points to this repository, and not your fork. If you're not sure what your remotes are, run `git remote -vvv`, there should be your fork and the holy/reference/base/origin/whatever-you-call-it repository.
2. A window will show up with many lines, replace `pick` with `fixup` on every line but the first one
3. Close your editor, git should do its magic, and you should end up with one commit
4. Use `git push --force` to overwrite what you already pushed. Don't forget the `--force` option otherwise git will try to merge both things together.

File: src/Resources/views/Admin/Security/login.html.twig [hunk 6]
```diff
@@ -49,36 +49,48 @@ file that was distributed with this source code.
...
+                {% else %}
+                    <form action="{{ path("sonata_user_admin_security_check") }}" method="post" role="form">
+                        <input type="hidden" name="_csrf_token" value="{{ csrf_token }}"/>
 
-                    <div class="form-group has-feedback">
-                        <input type="password" class="form-control" id="password" name="_password" required="required" placeholder="{{ 'security.login.password'|trans({}, 'SonataUserBundle') }}"/>
-                        <span class="glyphicon glyphicon-lock form-control-feedback"></span>
-                    </div>
+                        <div class="form-group has-feedback">
+                            <input type="text" class="form-control" id="username"  name="_username" value="{{ last_username }}" required="required" placeholder="{{ 'security.login.username'|trans({}, 'SonataUserBundle') }}"/>
```
Comment (greg0ire) [inline src/Resources/views/Admin/Security/login.html.twig]: Please linebreak this line.

File: src/Resources/views/Admin/Security/login.html.twig [hunk 7]
```diff
@@ -49,36 +49,48 @@ file that was distributed with this source code.
...
-                    <div class="row">
-                        <div class="col-xs-8">
-                            <div class="checkbox">
-                                <label>
-                                    <input type="checkbox" id="remember_me" name="_remember_me" value="on"/>
-                                    {{ 'security.login.remember_me'|trans({}, 'FOSUserBundle') }}
-                                </label>
-                            </div>
+                        <div class="form-group has-feedback">
+                            <input type="password" class="form-control" id="password" name="_password" required="required" placeholder="{{ 'security.login.password'|trans({}, 'SonataUserBundle') }}"/>
```
Comment (greg0ire) [inline src/Resources/views/Admin/Security/login.html.twig]: please linebreak this line.

File: src/Resources/views/Admin/Security/login.html.twig [hunk 8]
```diff
@@ -49,36 +49,48 @@ file that was distributed with this source code.
...
+                                <button type="submit" class="btn btn-primary btn-block btn-flat">{{ 'security.login.submit'|trans({}, 'FOSUserBundle') }}</button>
+                            </div>
                         </div>
-                    </div>
-                </form>
+                    </form>
 
-                {#<a href="{{ path('sonata_user_admin_resetting_request') }}">{{ 'forgotten_password'|trans({}, 'SonataUserBundle') }}</a>#}
-                <a href="{{ reset_route }}">{{ 'forgotten_password'|trans({}, 'SonataUserBundle') }}</a>
+                    {#<a href="{{ path('sonata_user_admin_resetting_request') }}">{{ 'forgotten_password'|trans({}, 'SonataUserBundle') }}</a>#}
```
Comment (greg0ire) [inline src/Resources/views/Admin/Security/login.html.twig]: Please remove commented code

File: src/Resources/views/Admin/Security/login.html.twig [hunk 9]
```diff
@@ -49,36 +49,49 @@ file that was distributed with this source code.
...
+                            </p>
+                        </div>
                     </div>
+                {% else %}
+                    <form action="{{ path("sonata_user_admin_security_check") }}" method="post" role="form">
+                        <input type="hidden" name="_csrf_token" value="{{ csrf_token }}"/>
+
+                        <div class="form-group has-feedback">
+                            <input type="text" class="form-control" id="username"  name="_username" value="{{ last_username }}"
+                             required="required" placeholder="{{ 'security.login.username'|trans({}, 'SonataUserBundle') }}"/>
```
Comment (greg0ire) [inline src/Resources/views/Admin/Security/login.html.twig]: Please use one attribute per line, and squash your commits

File: docs/reference/two_step_validation.rst [hunk 10]
```diff
@@ -35,5 +35,38 @@ Edit the configuration file:
         google_authenticator:
             enabled: true
             server:  yourserver.com
+            ip_white_list:
+                - 127.0.0.1
+            forced_for_role:
+                - ROLE_ADMIN
+
+Also, if you want to use ``ip_white_list`` and ``forced_for_role`` configuration nodes for automatically setting the secret
```
Comment (greg0ire) [inline docs/reference/two_step_validation.rst]: Please comply with the sf docs standards:

> [Each line should break approximately after the first word that crosses the 72nd character (so most lines end up being 72-78 characters);](http://symfony.com/doc/current/contributing/documentation/standards.html)

File: src/Resources/views/Admin/Security/login.html.twig [hunk 11]
```diff
@@ -49,36 +49,58 @@ file that was distributed with this source code.
...
+                        <input type="hidden" name="_csrf_token" value="{{ csrf_token }}"/>
+
+                        <div class="form-group has-feedback">
+                            <input type="text"
+                            class="form-control"
+                            id="username"
+                            name="_username"
+                            value="{{ last_username }}"
+                            required="required"
+                            placeholder="{{ 'security.login.username'|trans({}, 'SonataUserBundle') }}"/>
```
Comment (greg0ire) [inline src/Resources/views/Admin/Security/login.html.twig]: Almost good! Please indent all attributes, like this:
```html
<tag attr1="foo"
    attr2="bar"
    attr3="baz"
```

File: src/EventListener/TwoFactorLoginSuccessHandler.php [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+use Sonata\UserBundle\Model\User;
+use Sonata\UserBundle\Model\UserManagerInterface;
+use Symfony\Bundle\FrameworkBundle\Templating\EngineInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Authorization\AuthorizationCheckerInterface;
+use Symfony\Component\Security\Http\Authentication\AuthenticationSuccessHandlerInterface;
+
+class TwoFactorLoginSuccessHandler implements AuthenticationSuccessHandlerInterface
```
Comment (core23) [inline src/EventListener/TwoFactorLoginSuccessHandler.php]: final?

File: src/EventListener/TwoFactorLoginSuccessHandler.php [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+    ) {
+        $this->authorizationChecker = $authorizationChecker;
+        $this->engine = $engine;
+        $this->googleAuthenticator = $helper;
+        $this->userManager = $userManager;
+        $this->forcedForRoles = $forcedForRole;
+        $this->ipWhiteList = $ipWhiteList;
+    }
+
+    public function onAuthenticationSuccess(Request $request, TokenInterface $token)
```
Comment (core23) [inline src/EventListener/TwoFactorLoginSuccessHandler.php]: <s>Please add a PHPDoc with the return type unless we can't use return types in the method signature</s>

Please add the return type to the signature

File: src/Resources/config/google_authenticator.xml [hunk 14]
```diff
@@ -18,5 +18,13 @@
             <argument type="service" id="security.token_storage"/>
             <argument type="service" id="templating"/>
         </service>
+        <service id="sonata.user.google.authenticator.success_handler" class="Sonata\UserBundle\EventListener\TwoFactorLoginSuccessHandler">
```
Comment (core23) [inline src/Resources/config/google_authenticator.xml]: private or public?

File: src/EventListener/TwoFactorLoginSuccessHandler.php (hunk 13 above)
Comment (aLkRicha) [reply src/EventListener/TwoFactorLoginSuccessHandler.php]: @core23 t what method?

File: src/EventListener/TwoFactorLoginSuccessHandler.php (hunk 12 above)
Comment (aLkRicha) [reply src/EventListener/TwoFactorLoginSuccessHandler.php]: @core23 i'm not sure if this class needs to be final

File: src/EventListener/TwoFactorLoginSuccessHandler.php (hunk 13 above)
Comment (aLkRicha) [reply src/EventListener/TwoFactorLoginSuccessHandler.php]: @core23 add to both

File: src/EventListener/TwoFactorLoginSuccessHandler.php (hunk 12 above)
Comment (core23) [reply src/EventListener/TwoFactorLoginSuccessHandler.php]: We should make everything final by default and lower this if there are some good reasons. This allow us to make some changes without breaking BC.

Comment (aLkRicha) [reply src/EventListener/TwoFactorLoginSuccessHandler.php]: @core23 ok, done.

File: src/DependencyInjection/Configuration.php [hunk 15]
```diff
@@ -61,6 +61,16 @@ public function getConfigTreeBuilder()
...
+                            ->info('IPs for which 2FA will be skipped.')
+                        ->end()
+                        ->arrayNode('forced_for_role')
+                            ->prototype('scalar')->end()
+                            ->defaultValue(['ROLE_ADMIN'])
+                            ->info('User roles for which 2FA is necessarily.')
```
Comment (greg0ire) [inline src/DependencyInjection/Configuration.php]: "necessary"

**Comment by aLkRicha** [comment]
@greg0ire @core23 is it ok now?

File: src/EventListener/TwoFactorLoginSuccessHandler.php [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+use Sonata\UserBundle\Model\UserManagerInterface;
+use Symfony\Bundle\FrameworkBundle\Templating\EngineInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Authorization\AuthorizationCheckerInterface;
+use Symfony\Component\Security\Http\Authentication\AuthenticationSuccessHandlerInterface;
+
+final class TwoFactorLoginSuccessHandler implements AuthenticationSuccessHandlerInterface
```
Comment (OskarStark) [inline src/EventListener/TwoFactorLoginSuccessHandler.php]: please add `@author` tag

File: src/EventListener/TwoFactorLoginSuccessHandler.php [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+     */
+    private $forcedForRoles;
+
+    /**
+     * @var array
+     */
+    private $ipWhiteList;
+
+    /**
+     * AuthenticationSuccessHandler constructor.
```
Comment (OskarStark) [inline src/EventListener/TwoFactorLoginSuccessHandler.php]: please remove this comment

File: src/EventListener/TwoFactorLoginSuccessHandler.php [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+
+        if ($needToHave2FA && !$user->getTwoStepVerificationCode()) {
+            $secret = $this->googleAuthenticator->generateSecret();
+            $user->setTwoStepVerificationCode($secret);
+
+            $qrCodeUrl = $this->googleAuthenticator->getUrl($user);
+            $this->userManager->updateUser($user);
+
+            return $this->engine->renderResponse(
+                '@SonataUser/Admin/Security/login.html.twig',
```
Comment (OskarStark) [inline src/EventListener/TwoFactorLoginSuccessHandler.php]: is it correct to use hardcoded template paths here, or do we need to ask the pool for the configured templates?

@sonata-project/contributors

File: src/Resources/translations/SonataUserBundle.en.xliff [hunk 19]
```diff
@@ -554,6 +554,14 @@
                 <source>title_user_resetting_reset</source>
                 <target>Reset password</target>
             </trans-unit>
+            <trans-unit id="qr_code_secret">
+                <source>qr_code_secret</source>
+                <target>QR-code secret</target>
```
Comment (OskarStark) [inline src/Resources/translations/SonataUserBundle.en.xliff]: german translation: `QR-Code Secret`

File: src/Resources/views/Admin/Security/login.html.twig [hunk 20]
```diff
@@ -49,36 +49,58 @@ file that was distributed with this source code.
...
+                        <input type="hidden" name="_csrf_token" value="{{ csrf_token }}"/>
+
+                        <div class="form-group has-feedback">
+                            <input class="form-control"
+                            id="username"
+                            name="_username"
+                            placeholder="{{ 'security.login.username'|trans({}, 'SonataUserBundle') }}"
+                            required="required"
+                            type="text"
+                            value="{{ last_username }}"/>
```
Comment (greg0ire) [inline src/Resources/views/Admin/Security/login.html.twig]: Please indent lines 68 through 73

File: src/Resources/translations/SonataUserBundle.en.xliff [hunk 21]
```diff
@@ -554,6 +554,14 @@
...
+            </trans-unit>
+            <trans-unit id="qr_go_next">
+                <source>qr_go_next</source>
+                <target>I scanned it. Move on.</target>
```
Comment (OskarStark) [inline src/Resources/translations/SonataUserBundle.en.xliff]: german translation: `Ist gescannt. Weiter gehts.`

File: src/Resources/views/Admin/Security/login.html.twig [hunk 22]
```diff
@@ -49,36 +49,58 @@ file that was distributed with this source code.
...
-                        <input type="password" class="form-control" id="password" name="_password" required="required" placeholder="{{ 'security.login.password'|trans({}, 'SonataUserBundle') }}"/>
-                        <span class="glyphicon glyphicon-lock form-control-feedback"></span>
-                    </div>
+                        <div class="form-group has-feedback">
+                            <input class="form-control"
+                            id="password"
+                            name="_password"
+                            placeholder="{{ 'security.login.password'|trans({}, 'SonataUserBundle') }}"
+                            required="required"
+                            type="password"/>
```
Comment (greg0ire) [inline src/Resources/views/Admin/Security/login.html.twig]: Same here

File: src/Resources/translations/SonataUserBundle.en.xliff (hunk 19 above)
Comment (aLkRicha) [reply src/Resources/translations/SonataUserBundle.en.xliff]: @OskarStark done.

File: src/Resources/translations/SonataUserBundle.en.xliff (hunk 21 above)
Comment (aLkRicha) [reply src/Resources/translations/SonataUserBundle.en.xliff]: @OskarStark done.

**Comment by greg0ire** [comment]
>  Coverage decreased (-1.8%)

Isn't that a bit much?

**Comment by OskarStark** [comment]
there are no tests for `TwoFactorLoginSuccessHandler ` and it is much code

File: tests/EventListener/TwoFactorLoginSuccessHandlerTest.php [hunk 23]
```diff
@@ -0,0 +1,132 @@
...
+            $userManagerMock,
+            $forcedRoles,
+            $ipWhiteList
+        );
+        $this->request = Request::create('/');
+        if ($remoteAddr) {
+            $this->request->server->set('REMOTE_ADDR', $remoteAddr);
+        }
+        if ($needSession) {
+            $sessionMock = $this->getMockBuilder(SessionInterface::class)->getMock();
```
Comment (greg0ire) [inline tests/EventListener/TwoFactorLoginSuccessHandlerTest.php]: Please use `createMock`

File: tests/EventListener/TwoFactorLoginSuccessHandlerTest.php [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+    private function createTestClass($secret, $userRole, $remoteAddr, $needSession): void
+    {
+        $this->user = new BaseUser();
+        if ($secret) {
+            $this->user->setTwoStepVerificationCode($secret);
+        }
+        $this->token = new UsernamePasswordToken($this->user, null, 'admin', [new Role($userRole)]);
+        $tokenStorage = new TokenStorage();
+        $tokenStorage->setToken($this->token);
+        $authManagerMock = $this->getMockBuilder(AuthenticationManagerInterface::class)->getMock();
```
Comment (greg0ire) [inline tests/EventListener/TwoFactorLoginSuccessHandlerTest.php]: Please use `createMock`

File: tests/EventListener/TwoFactorLoginSuccessHandlerTest.php [hunk 25]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    /**
+     * @dataProvider data
+     *
+     * @param $secret
+     * @param $role
+     * @param $ip
+     * @param $needSession
+     * @param $expected
```
Comment (greg0ire) [inline tests/EventListener/TwoFactorLoginSuccessHandlerTest.php]: Please remove all the `@param` lines

File: tests/EventListener/TwoFactorLoginSuccessHandlerTest.php [hunk 26]
```diff
@@ -0,0 +1,132 @@
...
+     */
+    public function testDifferentCases($secret, $role, $ip, $needSession, $expected): void
```
Comment (greg0ire) [inline tests/EventListener/TwoFactorLoginSuccessHandlerTest.php]: Please add type hints

**Comment by greg0ire** [comment]
> Coverage increased (+3.0%)

Great job!!!

File: src/Resources/views/Admin/Security/login.html.twig [hunk 27]
```diff
@@ -49,36 +49,76 @@ file that was distributed with this source code.
...
+                    <div class="row">
+                        <div class="col-xs-12">
+                            <p align="middle"> <img src="{{ qrCodeUrl }}" alt="QR-code"></p>
+                            <p align="middle">{{ 'qr_code_secret'|trans({}, 'SonataUserBundle') }}: {{ qrSecret }}</p>
+                            <p align="middle">
+                                <a href="{{ path('sonata_admin_dashboard') }}">{{ 'qr_go_next'|trans({}, 'SonataUserBundle') }}</a>
+                            </p>
+                        </div>
                     </div>
+                {% elseif two_step_submit is defined and two_step_submit == true %}
```
Comment (greg0ire) [inline src/Resources/views/Admin/Security/login.html.twig]: `elseif two_step_submit|default(false)`

File: tests/EventListener/TwoFactorLoginSuccessHandlerTest.php [hunk 28]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+    public function data()
+    {
+        return [
+            [false, 'ROLE_USER', '192.168.1.1', false, Response::class],
+            [false, 'ROLE_ADMIN', false, false, RedirectResponse::class],
+            ['AQAOXT322JDYRKVJ', 'ROLE_ADMIN', '192.168.1.1', true, RedirectResponse::class],
+            [false, 'ROLE_ADMIN', '192.168.1.1', false, Response::class],
+
```
Comment (greg0ire) [inline tests/EventListener/TwoFactorLoginSuccessHandlerTest.php]: Please remove this line

**Comment by jordisala1991** [comment]
LGTM, Thank you @aLkRicha

**Pull request merged**